src/ligero/verifier.rs

```diff
@@ -1,3 +1,273 @@
+        // witnesses.
+        // Not documented in the specification.
+        .skip(tableau_layout.num_requested_columns())
+        .take(tableau_layout.witnesses_per_row())
+        .fold(FE::ZERO, |sum, term| sum + term);
+    if want_dot_product != proof_dot_product {
+        return Err(anyhow!("dot product mismatch"));
+    }
+
+    let inner_product_vector = inner_product_vector(
+        tableau_layout,
+        linear_constraints,
+        &linear_constraint_alphas,
+        quadratic_constraints,
+        &quad_constraint_alphas,
+    )?;
+
+    // Check that dot proof matches requested columns
+    let mut want_dot_row = proof.tableau_columns[TableauLayout::dot_proof_row()].clone();
+    let mut inner_product_vector_extended = Vec::with_capacity(tableau_layout.block_size());
+    for (products, tableau_row) in inner_product_vector
+        .chunks(tableau_layout.witnesses_per_row())
+        .zip(&proof.tableau_columns[tableau_layout.first_witness_row()..])
+    {
+        inner_product_vector_extended.truncate(0);
+        inner_product_vector_extended.resize(tableau_layout.num_requested_columns(), FE::ZERO);
+        inner_product_vector_extended.extend(products);
+
+        let extended = extend(&inner_product_vector_extended, tableau_layout.num_columns());
+        for ((want_dot_row_element, inner_product_element), tableau_element) in want_dot_row
+            .iter_mut()
+            .zip(tableau_layout.gather_iter(&extended, &requested_column_indices))
+            .zip(tableau_row)
+        {
+            *want_dot_row_element += inner_product_element * tableau_element;
+        }
+    }
+
+    let proof_dot_row = tableau_layout.gather(
+        &extend(&proof.dot_proof, tableau_layout.num_columns()),
+        &requested_column_indices,
+    );
+
+    if want_dot_row != proof_dot_row {
+        return Err(anyhow!("dot proof mismatch"));
+    }
+
+    // Check that low degree test proof matches
+    let mut want_low_degree_row =
+        proof.tableau_columns[TableauLayout::low_degree_test_row()].clone();
+
+    for (proof_column, challenge) in proof
+        .tableau_columns
+        .iter()
+        .skip(tableau_layout.first_witness_row())
+        .zip(low_degree_test_blind)
+    {
+        for (ldt_column, witness_column) in want_low_degree_row.iter_mut().zip(proof_column.iter())
+        {
+            *ldt_column += challenge * witness_column;
+        }
+    }
+
+    let proof_low_degree_test_row = tableau_layout.gather(
+        &extend(&proof.low_degree_test_proof, tableau_layout.num_columns()),
+        &requested_column_indices,
+    );
+
+    if want_low_degree_row != proof_low_degree_test_row {
+        return Err(anyhow!("low degree test proof mismatch"));
+    }
+
+    // Check that the quadratic proof matches
+    let mut want_quadratic_test_row =
+        proof.tableau_columns[TableauLayout::quadratic_test_row()].clone();
+    assert_eq!(
```

Comment:
We wouldn't want this to panic in production code, right? Like, this is untrusted input, so badness could ensue here quite easily.